### ISSUE TITLE
Request to add CMake build scripts

### ISSUE THREAD

**Issue by Type1J** [nghttp2/nghttp2#448]
Using and debugging apps and libnghttp2 itself would be much easier if it had a CMake build system. Please consider adding a CMakeLists.txt that works for at least Windows, OSX, and Linux.

**Comment by tatsuhiro-t** [comment]
I will happily accept CMake support PR.

**Comment by Lekensteyn** [comment]
FYI I have started with adding CMake support. WIP at https://github.com/Lekensteyn/nghttp2/compare/master...cmake

**Comment by tatsuhiro-t** [comment]
Great!

**Comment by Type1J** [comment]
@Lekensteyn Thanks! This will really help me out.

**Comment by wrowe** [comment]
@Lekensteyn you probably have this covered but there was a minimal implementation that might be helpful if you wanted to look;

**Comment by Lekensteyn** [comment]
@wrowe Were you referring to https://github.com/trawick/nghttp2-minimal-cmake? At this moment of writing, that file is providing support to compile objects, but missing configure checks while my work has exactly the opposite problem (it implements most configure checks, but lacks sources for compilation).

I was considering to split the configure checks (~500 lines) and sources (0 lines) into separate cmake files, but maybe it's better to go ahead and finish an initial version that can actually compile stuff.

**Comment by wrowe** [comment]
Yes @Lekensteyn - was just pointing this out for background.  It might be easiest time to follow Jeff's model and get the underlying lib built, then introducing the client, and server.  I'm happy to test and hopefully contribute to the effort, Trawick may be interested as well.

**Comment by Lekensteyn** [comment]
Some success! I can compile and install a library and pkg-config file:

    mkdir build && cd build
    cmake .. -DENABLE_WERROR=1
    make install DESTDIR=fs

Output for `make install DESTDIR=fs`:
```
Scanning dependencies of target nghttp2
[  4%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_map.c.o
[  9%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_frame.c.o
[ 14%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_queue.c.o
[ 19%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_pq.c.o
[ 23%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_buf.c.o
[ 28%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_stream.c.o
[ 33%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_outbound_item.c.o
[ 38%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_session.c.o
[ 42%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_submit.c.o
[ 47%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_helper.c.o
[ 52%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_npn.c.o
[ 57%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_hd_huffman.c.o
[ 61%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_hd.c.o
[ 66%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_hd_huffman_data.c.o
[ 71%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_version.c.o
[ 80%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_priority_spec.c.o
[ 80%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_option.c.o
[ 85%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_callbacks.c.o
[ 90%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_mem.c.o
[ 95%] Building C object lib/CMakeFiles/nghttp2.dir/nghttp2_http.c.o
[100%] Linking C shared library libnghttp2.so
[100%] Built target nghttp2
Install the project...
-- Install configuration: ""
-- Installing: fs/usr/local/lib64/libnghttp2.so.18.1.4
-- Installing: fs/usr/local/lib64/libnghttp2.so.18
-- Installing: fs/usr/local/lib64/libnghttp2.so
-- Installing: fs/usr/local/lib64/pkgconfig/libnghttp2.pc
```

The .pc file looks sane, I have not tried to use the library yet. While I tried to define all macros as needed (grepped for `HAVE_[A-Z0-9_]+_H` and similar constructs without `_H`), it is still possible that I overlooked something. Next target is `src/`!

**Comment by alagoutte** [comment]
Do not forget to add step on .travis.yml (you can look travis.yml from @wireshark)

**Comment by Lekensteyn** [comment]
@alagoutte Will do that once the patches are somehow mergeable.

third-party and src also compile now! Tested with:
```
cmake .. -DENABLE_WERROR=1 -DENABLE_APP=1 -DENABLE_HPACK_TOOLS=1 -DWITH_LIBXML2=1 -DWITH_JEMALLOC=1 && make
```

@tatsuhiro-t Would you prefer a single, squashed patch at the end of the series, or the tree with cmake changes as-is?

**Comment by tatsuhiro-t** [comment]
I'm ok either way.  Let's use changes as-is.

**Comment by Lekensteyn** [comment]
Update since last time: the documentation and integration tests can also run. Examples are built too.

Remaining work to do:
 - [ ] integrate cunit and the tests directory
 - [ ] fix python bindings (cython or ccfi?)
 - [ ] improve package detection (consider cmake `find_package` instead of pkg-config because it works better with non-Linux environments)
 - [ ] Implement auto-detection of dependencies (currently optional features are off by default)
 - [ ] Integrate mruby and neverbleed
 - [ ] Visit all `XXX` markers in the cmake files
 - [ ] travis CI hook
 - [ ] maybe add license headers to files? (it this really needed? some files are only two lines...)
 - [ ] include CMakeLists.txt and related files in `EXTRA_DIST` (Makefile.am)

Maybe also consider using CPack for distribution packages, but as long as autotools is still in the repo, distribution tarballs need to be built anyway for all `configure`, `Makefile.in`, etc. files.

I have rebased the branch to drop some unrelated patches. (rebased on master + build-fixes)

**Comment by Lekensteyn** [comment]
The CMake build is very usable now. I have only tested it on Arch Linux with CMake 3.4 now, please check whether it works for you! Patches for (the remaining) issues are also welcome :-)

Basic test (auto-detection of packages should work):

    mkdir build && cmake ..

This is what I have been using to test all of the features (I built spdylay in /tmp/spdylay and installed it under subdir root/, but you can also set `SPDYLAY_LIBRARY` and `SPDYLAY_INCLUDE_DIR` instead of using pkg-config; you will also have to `git submodule sync` for mruby and neverbleed):

    time PKG_CONFIG_PATH=/tmp/spdylay/root/lib/pkgconfig \
    cmake ~/apps/nghttp2 -GNinja \
    -DWITH_MRUBY=1 -DWITH_NEVERBLEED=1 -DENABLE_ASIO_LIB=ON  \
    -DCUNIT_LIBRARY=/tmp/cunit/pkg/cunit/usr/lib/libcunit.so \
    -DCUNIT_INCLUDE_DIR=/tmp/cunit/pkg/cunit/usr/include
    time ninja

**Comment by tatsuhiro-t** [comment]
Awesome!  Will check your work.  Thank you.

**Comment by tatsuhiro-t** [comment]
Sorry for late response. I tested cmake branch.  After git clone, what I did:
```
$ git submodule update --init
$ cmake . -DWITH_MRUBY=1 -DWITH_NEVERBLEED=1
```

And compile started.  Then I got the following error:
```
[ 85%] Building CXX object src/CMakeFiles/nghttpx_static.dir/shrpx_mruby_module_request.cc.o
[ 86%] Building CXX object src/CMakeFiles/nghttpx_static.dir/shrpx_mruby_module_response.cc.o
[ 88%] Linking CXX static library libnghttpx.a
[ 88%] Built target nghttpx_static
Scanning dependencies of target nghttpx
[ 89%] Building CXX object src/CMakeFiles/nghttpx.dir/shrpx.cc.o
make[2]: *** No rule to make target 'third-party/mruby/build/lib/libmruby.a', needed by 'src/nghttpx'.  STOP.
CMakeFiles/Makefile2:532: recipe for target 'src/CMakeFiles/nghttpx.dir/all' failed
make[1]: *** [src/CMakeFiles/nghttpx.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
```

Perhaps, we should build libmruby.a before building nghttpx?

**Comment by Lekensteyn** [comment]
I've fixed the mruby build in Lekensteyn/nghttp2@d6c3b6e84ca8dde0d8a130440e831d3dc89258eb and also updated the gitignore file. (mruby will always be considered "out of date" and invoked when building)

**Comment by tatsuhiro-t** [comment]
That fixed my problem.  Thank you!
`make it` works great.

I tried to run `make test` but it failed.  After running `make main`, `make failmalloc`, and `make nghttpx-unittest`, `make test` worked.

**Comment by Lekensteyn** [comment]
"test" does not build tests [by default][1], I have added a "check" target now that builds tests before invoking "test". Also added stuff to EXTRA_DIST

 [1]: https://stackoverflow.com/questions/733475/cmake-ctest-make-test-doesnt-build-tests

**Comment by tatsuhiro-t** [comment]
Fixed with #544.  Thank you!
